src/solvers/strings/array_pool.cpp

```diff
@@ -115,6 +115,54 @@ array_string_exprt array_poolt::make_char_array_for_char_pointer(
   return to_array_string_expr(insert_result.first->second);
 }
 
+/// Given an array_string_exprt, get the size of the underlying array. If that
+/// size is undefined, create a new symbol for the size.
+/// Then add an entry from `array_expr` to that size in the `length_of_array`
+/// map.
+///
+/// If `array_expr` is an `if_exprt`, recursively add entries for the parts of
+/// the `if_exprt`, but not for the `if_exprt`s themselves.
+///
+/// This function should only be used at the creation of the
+/// `array_string_exprt`s, as it is the only place where we can reliably refer
+/// to the size in the type of the array.
+static void attempt_assign_length_from_type(
+  const array_string_exprt &array_expr,
+  std::unordered_map<array_string_exprt, exprt, irep_hash> &length_of_array,
+  symbol_generatort &symbol_generator)
+{
+  if(const auto &if_expr = expr_try_dynamic_cast<if_exprt>((exprt)array_expr))
+  {
+    attempt_assign_length_from_type(
+      to_array_string_expr(if_expr->true_case()),
+      length_of_array,
+      symbol_generator);
+    attempt_assign_length_from_type(
+      to_array_string_expr(if_expr->false_case()),
+      length_of_array,
+      symbol_generator);
+  }
+  else
+  {
+    const exprt &size_from_type = to_array_type(array_expr.type()).size();
+    if(size_from_type != infinity_exprt(size_from_type.type()))
+    {
+      const auto emplace_result =
+        length_of_array.emplace(array_expr, size_from_type);
+      INVARIANT(
+        emplace_result.second,
+        "attempt_assign_length should only be called when no entry for the"
+        "array_string_exprt exists in the length_of_array map");
+    }
+    else
+    {
+      length_of_array.emplace(
```

Comment:
Should check this one is new, too? Therefore, deduplicate the `emplace` calls and the invariant below this if block?

Reply:
Done in fixup commit